### PR TITLE
Update order cancellation, sets inventory units to canceled

### LINKED ISSUE #3719: Inventory units stay `on_hand` after an order is canceled
**Is your feature request related to a problem? Please describe.**

When I cancel and order, the inventory units remain `on_hand`
```
# Canceled order
order.inventory_units.pluck(:state) => ["on_hand"]
```
**Steps to recreate**
- Create an order and complete it
- Cancel the order by clicking the "Cancel" button
- Notice that the order state is `canceled`, the shipment state it `canceled`, but the inventory units states remain `on_hand` (see screenshot below)

**Describe the solution you'd like**

When I cancel an order, the inventory units state should also change to `canceled`
```
# Desired return for canceled order
order.inventory_units.pluck(:state) => ["canceled"]`
```

**Describe alternatives you've considered**

**Additional context**

The issue we've specifically had has been with our warehouse system. The inventory units _have_ to be canceled so our CSRs have to cancel and order, _and_ cancel all the items under the "Cancel Items" tab. But it does seem strange that when you cancel an order the inventory units are not also canceled. We set the shipment state to cancelled so I would assume we do the same for the inventory units. 
cancelled-order

### PR DESCRIPTION
**Pull request by stefano-sarioli**
This PR fix #3719.

Before this commit when an order was canceled the inventory units
of the order will remain untouched. So there were canceled orders with
inventory units with the state on_hand or backordered.

This situation was not only inconsistent but can lead to subtle bugs
with the stock management system.
How to reproduce the bug: 

- Create an order for a single variant, but select a number of items greater than the available amount.
- The order generates 2 shipments, one with the inventory units on_hand and one with the inventory units back-ordered.
- Cancel the order.
- The inventory units on_hand remains on_hand.
- The inventory units back-ordered go to the on_hand state. 
- If you resume the order all the inventory units will be on_hand, and the shipment will be ready to be shipped.

This problem is caused by the fact that when the order is canceled, a movement is created to restock the stock_item correlated with the inventory_units. As soon as the stock_items get back to a positive value it will try to convert as much as possible back-ordered inventory units to on_hand.

order_complete_before
order_canceled_before
order_resumed_before

With this commit the inventory units related to an order are canceled when the order is canceled.

When an order is resumed we recalculate the state of the connected
inventory units, based on the actual stocks available.

order_complete_after
order_canceled_after
order_resumed_after

What do you think about this issue, and my solution?
I'm not sure that the testing is enough, probably I can add an E2E test, what do you think? any idea on what can be the best place to put this test?

**Checklist:**
- [X] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [X] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [X] I have added tests to cover this change (if needed)
- [X] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: core/app/models/spree/shipment.rb
Comment (kennyadsl) [inline core/app/models/spree/shipment.rb:397]: Is canceling an order the only way to cancel an IU at the moment? Just wondering if other IUs can have the state canceled and they would be resumed, even if not part of the order cancelation.

**Review by kennyadsl — COMMENTED** [review]
Left a preliminary concern, also some specs seem to be failing.

File: core/app/models/spree/shipment.rb
Comment (stefano-sarioli) [reply core/app/models/spree/shipment.rb:397]: No, you are right, there are other places.
I have to pick only the canceled units for the variant reported on the manifest items.

Comment (stefano-sarioli) [reply core/app/models/spree/shipment.rb]: I've used the line_item to be sure.

Comment (stefano-sarioli) [reply core/app/models/spree/shipment.rb]: BTW I'm double-checking this to be sure. Sorry for the message spam 🙇 .

File: core/app/models/spree/shipment.rb
Comment (kennyadsl) [inline core/app/models/spree/shipment.rb:400]: Are these variables filled with the current availability in the stock location, right?

File: core/spec/models/spree/shipment_spec.rb
Comment (kennyadsl) [inline core/spec/models/spree/shipment_spec.rb:503]: Thanks for adding this spec!

**Review by kennyadsl — COMMENTED** [review]
Just left a question, overall I can't see any issues with this approach.

File: core/spec/models/spree/shipment_spec.rb
Comment (stefano-sarioli) [reply core/spec/models/spree/shipment_spec.rb:503]: Thanks to you for pointing out the edge case 🙇, I totally missed it!

File: core/app/models/spree/shipment.rb
Comment (stefano-sarioli) [reply core/app/models/spree/shipment.rb:400]: Exactly, this will update the inventory units state based on the current stock level.

This is why inside #after_resume I have to call #resume_inventory_units before unstocking the items. If you call #resume_inventory_units after the unstocking the states of the inventory units will be wrong. 

There is a test covering this use-case to avoid regression.

**Review by kennyadsl — APPROVED** [review]
Thanks @stefano-sarioli !

File: core/lib/spree/core/state_machines/inventory_unit.rb
Comment (spaghetticode) [inline core/lib/spree/core/state_machines/inventory_unit.rb:38]: I'm not 100% sure about this event name. 

Semantically speaking, event names should be verbs, as they actively change the state of an object. In this case, the action name is the same as the target state, and this may lead to some confusion too.  
Also, there's the query method #can_on_hand? added automatically, which again does not make 100% sense to me. 

What about changing the name to something like `set_on_hand` or `make_on_hand`? And if you find a better naming, just go for it.

File: core/app/models/spree/shipment.rb
Comment (kennyadsl) [inline core/app/models/spree/shipment.rb:395]: We discussed this within the core team. We think that resuming concerns should be moved into a separate PR. There are a lot of opinions on how resuming an order should work and probably none of them will work as the majority of stores expect (everyone wants to restock items in different ways, depending on if they are using Solidus to manage stocks, which itself has a lot of variables, or they are relying on an external system). 

We are thinking to actually deprecate the Resume order functionality. It's not used a lot and gives us more headaches than the benefits it provides to stores (none of us can think of a single store that ever used this functionality). The way to go here is moving the current behavior into a legacy extension, and deprecate the functionality in core (removing the Resume button from the admin UI). WDYT?

Comment (stefano-sarioli) [reply core/app/models/spree/shipment.rb:395]: I agree with you, especially if it's an underused functionality. 
My solution is far too general and makes a lot of assumptions about the stock management system.

I think that extracting the resume functionality into an external gem, maybe with this fix on it, would be the best solution.

Probably we can close this PR, I'll wait for the final decision about the resume functionality and will come back to contribute to that one as soon as it's ready.

Thank a lot for your feedback and consideration 🙏.

Comment (kennyadsl) [reply core/app/models/spree/shipment.rb:395]: The cancellation part of this PR is well done and something that we definitely need in core. Please don't close the PR but just remove the part that resumes the inventory units. 🙏

File: core/lib/spree/core/state_machines/inventory_unit.rb
Comment (stefano-sarioli) [reply core/lib/spree/core/state_machines/inventory_unit.rb:38]: I agree with you, I was not convinced too by the name of that event but couldn't think of something better. 

I like `set_on_hand` proposed by you ❤️.

Sadly, as pointed out in the last comment of Kennyadsl, I'm probably going to close this PR to move the whole resume functionality out of Solidus.

Thanks a lot for your feedback, I'll note it down to use it when the work on the extracted functionality will start 🙏.

**Pull request closed without merging**
